Shapely functions as sample process

jeafreezy opened the pull request:
# Overview

This PR exposes some selected shapely functions as sample process. The selection cut across different operations in shapely. It uses the name of the category of that function as the namespace. E.g union operation under the set module is described as `set:union`. This is to avoid clashes with other functions.


As mentioned in the issue, it accepts list of geometry inputs (WKT and/or GeoJSON geometry), performs the operation and returns the result in the specified **output_format** (optional).

The approach is describe in the image below:

![image](https://github.com/geopython/pygeoapi/assets/36086324/2010d630-5623-4864-89c8-a6fbe28cffd1)


# Related issue / discussion

https://github.com/geopython/pygeoapi/issues/1439

# Additional information

**Todo/Idea/Challenge**

- Tests(units) - I didn't see any test for process in the codebase. I tried writing one, but I need to spend some more time to understand the test structure.

- I am yet to figure out how to pass optional function parameters in the inputs. I am still confused on the data type to use in the metadata, especially since different functions require different parameters. e.g **buffer** operation requires **distance**. @tomkralidis and I discussed an alternative approach earlier today, which is using different classes for different functions. That could potentially solve this issue. But I didn't have much time to explore that. Although this may also lead to repetitive code. Since the idea is to have it as a sample, hopefully this current implementation is sufficient.

- Idea: Perhaps in the future, it will be great to chain functions. E.g passing multiple **operations** like ['buffer', 'join'] can be interpreted as perfoming the **buffer** operation on the geometries, followed by the **join** operation before returning the result.

# Dependency policy (RFC2)

- [x] I have ensured that this PR meets [RFC2](https://pygeoapi.io/development/rfc/2) requirements

# Updates to public demo

- [ ] I have ensured that breaking changes to the [pygeoapi master demo server](https://github.com/geopython/demo.pygeoapi.io) have been addressed
  - [ ] https://github.com/geopython/demo.pygeoapi.io/blob/master/services/pygeoapi_master/local.config.yml

# Contributions and licensing

(as per https://github.com/geopython/pygeoapi/blob/master/CONTRIBUTING.md#contributions-and-licensing)

- [x] I'd like to contribute [feature X|bugfix Y|docs|something else] to pygeoapi. I confirm that my contributions to pygeoapi will be compatible with the pygeoapi license guidelines at the time of contribution
- [x] I have already previously agreed to the pygeoapi Contributions and Licensing Guidelines

francbartoli commented:
@jeafreezy thanks for your great contributions. I have just one general comment about security:
- Exposing these functions publicly would leave a big surface for DoS attacks so we should make sure that they are not enabled as default processes sample
- My suggestion is to pick up them from the configuration if additionally configured

tomkralidis commented:
Agree; it makes sense to have the plugin in the codebase without having in default configuration.

jeafreezy commented:
Thank you both for the review and feedback. Please, how do you think I should go about this? I have a few ideas;

- Should there be a new entry in the **example-config.yml** for functions, where the user can list the functions they want to enable? Something like below;

![image](https://github.com/geopython/pygeoapi/assets/36086324/5c15c53d-833d-4594-8ad2-42d7225ef475)

- Or the shapely-functions entry should be commented out in the config, so they user can enable it, if they want to ? Something like below;

![image](https://github.com/geopython/pygeoapi/assets/36086324/806c30ec-2a09-4ea1-b1e5-dbd95c575747)

- Or is there a better approach to implement the suggestion ?

Kindly advise. Thank you.

webb-ben commented:
Hi @jeafreezy, I believe the correct approach would be to omit the configuration entirely from the default pygeoapi configuration file, and instead, documenting the capacities of the Shapely Process Provider [here](https://docs.pygeoapi.io/en/latest/data-publishing/ogcapi-processes.html) / [location in source code](https://github.com/geopython/pygeoapi/blob/master/docs/source/data-publishing/ogcapi-processes.rst)

jeafreezy commented:
Thank you @webb-ben . I will work on that

webb-ben commented on .pre-commit-config.yaml:
```diff
@@ -4,16 +4,15 @@
 files: '\.(py|sh|sql)$'
```
Are changes to this file necessary?

jeafreezy replied:
![image](https://github.com/geopython/pygeoapi/assets/36086324/a662b9f0-a37a-478a-bb4d-91ab3dde8914)

I believe this happened when I synced with the main branch recently. I was not my edit.

webb-ben commented on docs/source/data-publishing/ogcapi-processes.rst:
```diff
@@ -9,17 +9,24 @@ fashion (inputs, outputs).
        hello-world:
            processor:
                name: HelloWorld
+        
+        #Add this block to the configuration file to enable shapely processes.
+        #shapely-functions:
+           #processor:
+               #name: ShapelyFunctions
```
Could you move this to the start of the Shapely section, with the process uncommented?
```md
Shapely Functions (Optional)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: yaml

        shapely-functions:
           processor:
               name: ShapelyFunctions
```

jeafreezy replied:
Alright. Will do.

webb-ben commented on pygeoapi-config.yml:
```diff
@@ -302,3 +302,4 @@ resources:
         type: process
```
Are changes to this file necessary?

jeafreezy replied:
It's as a result of syncing with the master branch.

tomkralidis replied:
Can we remove this addition?

webb-ben commented on pygeoapi/templates/tilematrixsets/tilematrixset.html:
```diff
@@ -1,18 +1,19 @@
-/ <a href="../TileMatrixSets/{{ data['id'] }}?f=html">{{ data['id'] }}</a>
-{% endblock %}
-{% block body %}
+{% extends "_base.html" %} {% block title %}{{ super() }} {% trans
```
Are changes to this file necessary?

jeafreezy replied:
I didn't changed the files. But I think it was just reformatted by my editor.

webb-ben reviewed (changes requested):
Not sure if we need need pytest to ensure function of the process capacities. 

Apart from that, small changes to docs and a couple of files I am not sure need to be included in this PR, got my +1.

jeafreezy commented:
> Not sure if we need need pytest to ensure function of the process capacities.
> 
> Apart from that, small changes to docs and a couple of files I am not sure need to be included in this PR, got my +1.

Thank you for reviewing my PR. I will address the issues.

tomkralidis commented on .gitignore:
Can we remove this addition?

tomkralidis commented on .pre-commit-config.yaml:
Can we remove this addition?

tomkralidis commented on docs/source/data-publishing/ogcapi-processes.rst:
```diff
@@ -9,14 +9,16 @@ fashion (inputs, outputs).
 allowing developers to implement custom processing workflows in Python.
 
-A `sample`_ ``hello-world`` process is provided with the pygeoapi default configuration.
+The pygeoapi offers two processes: a default ``hello-world`` process which allows you to quickly explore the capabilities of processes, and an optional ``shapely-functions`` process with more advanced features that leverages Shapely_ to expose various geometric processing functionality.
```
`Shapely_` -> `` `Shapely`_``

tomkralidis commented on docs/source/data-publishing/ogcapi-processes.rst:
```diff
@@ -107,14 +112,87 @@ Processing examples
+Shapely Functions (Optional)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The `shapely-functions` process exposes some selected Shapely_ functions as sample process. The selection cut across different operations in shapely. To avoid function collision, it uses the name of the function category as the namespace. E.g *union* operation under the *set* module is described as *set:union*.
```
- `Shapely_` -> `` `Shapely`_``
- shapely -> Shapely

tomkralidis commented on pygeoapi/process/shapely_functions.py:
```diff
@@ -0,0 +1,409 @@
+#
+# =================================================================
+
+import enum
```
Imports should be spaced and ordered as follows:

```python
import enum
import logging

from typing import Any, Dict, List, Tuple, Union

from shapely.geometry import mapping, shape
from shapely.geometry.base import BaseGeometry
from shapely.prepared import prep
from shapely.wkt import loads

from pygeoapi.process.base import BaseProcessor, ProcessorExecuteError
```

tomkralidis commented on pygeoapi/process/shapely_functions.py:
```diff
@@ -0,0 +1,409 @@
+            "type": "text/html",
+            "rel": "about",
+            "title": "information",
+            "href": "https://example.org/process",
```
Update link to Shapely, i.e. `https://shapely.readthedocs.io`

tomkralidis commented on pygeoapi/templates/tilematrixsets/tilematrixset.html:
```diff
@@ -15,4 +15,4 @@ <h1>{{ data['id'] }} tile matrix set</h1>
 
 </section>
-{% endblock %}
+{% endblock %}
```
Can we remove this addition?

tomkralidis commented on pygeoapi/process/shapely_functions.py:
Overall comment: any Python objects with trailing commas (dicts, lists, etc.) should have the trailing commas removed; example:

from

```python
dict_ = {
    'foo': 'bar',
}
```

to


```python
dict_ = {
    'foo': 'bar'
}
```

jeafreezy replied:
Thank you for the feedback @tomkralidis

jeafreezy commented:
Thank you for reviewing my PR @tomkralidis. I have addressed the issues to the best of my knowledge. Kindly let me know if there is anything else I missed. Thank you.
